### ISSUE TITLE
Pod Garbage collector fails to clean up  PODs from nodes that are not running anymore

### ISSUE THREAD

**Issue by carlosjgp** [kubernetes/kubernetes#118261]
### What happened?

This happened after upgrading to Kubernetes 1.26 from 1.24

The succession of events we have observed:
- Our workloads use HorizontalPodAutoscaler to scale up due to traffic increase
- Cluster Autoscaler provisions new nodes to accommodate the new replicas
- Traffic goes down
- HPA downsize replicas
- Cluster Autoscaler notices nodes under the threshold of usage and PODs can be accommodated into other nodes
- Nodes are drained and downscaled
- Kubernetes still reports PODs in a "Running" or "Terminating" state in the node that no longer exists
- Kubernetes control plane reports "Orphan pods" on its audit logs

### What did you expect to happen?

Kubernetes Garbage collector to clean up these PODs after the node is gone

### How can we reproduce it (as minimally and precisely as possible)?

Deploy the following yaml deployments into a K8s cluster 1.26 and terminate one of the nodes

(duplicated port key "containerPort + protocol")
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app.kubernetes.io/name: nginx
    app.kubernetes.io/instance: nginx
    app.kubernetes.io/component: deployment
spec:
  replicas: 100
  selector:
    matchLabels:
      app.kubernetes.io/name: nginx
      app.kubernetes.io/instance: nginx
      app.kubernetes.io/component: deployment
  template:
    metadata:
      labels:
        app.kubernetes.io/name: nginx
        app.kubernetes.io/instance: nginx
        app.kubernetes.io/component: deployment
    spec:
      restartPolicy: Always
      containers:
        - name: nginx
          image: "nginx:latest"
          imagePullPolicy: IfNotPresent
          ports:
            - containerPort: 8080
              name: http
              protocol: TCP
            - containerPort: 8080
              name: metrics
              protocol: TCP
```

(Duplicated environment variable. Set twice by mistake)
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-2
  labels:
    app.kubernetes.io/name: nginx-2
    app.kubernetes.io/instance: nginx-2
    app.kubernetes.io/component: deployment
spec:
  replicas: 100
  selector:
    matchLabels:
      app.kubernetes.io/name: nginx-2
      app.kubernetes.io/instance: nginx-2
      app.kubernetes.io/component: deployment
  template:
    metadata:
      labels:
        app.kubernetes.io/name: nginx-2
        app.kubernetes.io/instance: nginx-2
        app.kubernetes.io/component: deployment
    spec:
      restartPolicy: Always
      containers:
        - name: nginx-2
          image: "nginx:latest"
          imagePullPolicy: IfNotPresent
          env:
            - name: MY_VAR
              value: value-1
            - name: MY_VAR
              value: value-2
```

### Anything else we need to know?

(Duplicated port error)
```json
{
  "kind": "Event",
  "apiVersion": "audit.k8s.io/v1",
  "level": "RequestResponse",
  "auditID": "9d3c7dbf-f599-422b-866c-84d52f3b1a22",
  "stage": "ResponseComplete",
  "requestURI": "/api/v1/namespaces/app-b/pods/app-b-5894548cb-7tssd/status?fieldManager=PodGC\u0026force=true",
  "verb": "patch",
  "user":
    {
      "username": "system:serviceaccount:kube-system:pod-garbage-collector",
      "uid": "f099fed7-6a3d-4a3b-bc1b-49c668276d76",
      "groups":
        [
          "system:serviceaccounts",
          "system:serviceaccounts:kube-system",
          "system:authenticated",
        ],
    },
  "sourceIPs": ["172.16.38.214"],
  "userAgent": "kube-controller-manager/v1.26.4 (linux/amd64) kubernetes/4a34796/system:serviceaccount:kube-system:pod-garbage-collector",
  "objectRef":
    {
      "resource": "pods",
      "namespace": "app-b",
      "name": "app-b-5894548cb-7tssd",
      "apiVersion": "v1",
      "subresource": "status",
    },
  "responseStatus":
    {
      "metadata": {},
      "status": "Failure",
      "message": 'failed to create manager for existing fields: failed to convert new object (app-b/app-b-5894548cb-7tssd; /v1, Kind=Pod) to smd typed: .spec.containers[name="app-b"].ports: duplicate entries for key [containerPort=8082,protocol="TCP"]',
      "code": 500,
    },
  "requestObject":
    {
      "kind": "Pod",
      "apiVersion": "v1",
      "metadata":
        {
          "name": "app-b-5894548cb-7tssd",
          "namespace": "app-b",
        },
      "status":
        {
          "phase": "Failed",
          "conditions":
            [
              {
                "type": "DisruptionTarget",
                "status": "True",
                "lastTransitionTime": "2023-05-23T17:00:55Z",
                "reason": "DeletionByPodGC",
                "message": "PodGC: node no longer exists",
              },
            ],
        },
    },
  "responseObject":
    {
      "kind": "Status",
      "apiVersion": "v1",
      "metadata": {},
      "status": "Failure",
      "message": 'failed to create manager for existing fields: failed to convert new object (app-b/app-b-5894548cb-7tssd; /v1, Kind=Pod) to smd typed: .spec.containers[name="app-b"].ports: duplicate entries for key [containerPort=8082,protocol="TCP"]',
      "code": 500,
    },
  "requestReceivedTimestamp": "2023-05-23T17:00:55.648887Z",
  "stageTimestamp": "2023-05-23T17:00:55.652513Z",
  "annotations":
    {
      "authorization.k8s.io/decision": "allow",
      "authorization.k8s.io/reason": 'RBAC: allowed by ClusterRoleBinding "system:controller:pod-garbage-collector" of ClusterRole "system:controller:pod-garbage-collector" to ServiceAccount "pod-garbage-collector/kube-system"',
    },
}
```

(Duplicated env var error)
```json
{
  "kind": "Event",
  "apiVersion": "audit.k8s.io/v1",
  "level": "RequestResponse",
  "auditID": "9ffc9212-3d74-4b86-98bb-5e6f0c5395b1",
  "stage": "ResponseComplete",
  "requestURI": "/api/v1/namespaces/app-a/pods/app-a-7b7ddc5874-c85hq/status?fieldManager=PodGC\u0026force=true",
  "verb": "patch",
  "user":
    {
      "username": "system:serviceaccount:kube-system:pod-garbage-collector",
      "uid": "f099fed7-6a3d-4a3b-bc1b-49c668276d76",
      "groups":
        [
          "system:serviceaccounts",
          "system:serviceaccounts:kube-system",
          "system:authenticated",
        ],
    },
  "sourceIPs": ["172.16.38.214"],
  "userAgent": "kube-controller-manager/v1.26.4 (linux/amd64) kubernetes/4a34796/system:serviceaccount:kube-system:pod-garbage-collector",
  "objectRef":
    {
      "resource": "pods",
      "namespace": "app-a",
      "name": "app-a-7b7ddc5874-c85hq",
      "apiVersion": "v1",
      "subresource": "status",
    },
  "responseStatus":
    {
      "metadata": {},
      "status": "Failure",
      "message": "failed to create manager for existing fields: failed to convert new object (app-a/app-a-7b7ddc5874-c85hq; /v1, Kind=Pod) to smd typed: errors:\n .spec.containers[name=\"app-a\"].env: duplicate entries for key [name=\"RABBITMQ_HOST\"]\n .spec.containers[name=\"app-a\"].env: duplicate entries for key [name=\"RABBITMQ_PORT\"]\n .spec.initContainers[name=\"db-migration\"].env: duplicate entries for key [name=\"RABBITMQ_HOST\"]\n .spec.initContainers[name=\"db-migration\"].env: duplicate entries for key [name=\"RABBITMQ_PORT\"]",
      "code": 500,
    },
  "requestObject":
    {
      "kind": "Pod",
      "apiVersion": "v1",
      "metadata":
        {
          "name": "app-a-7b7ddc5874-c85hq",
          "namespace": "app-a",
        },
      "status":
        {
          "phase": "Failed",
          "conditions":
            [
              {
                "type": "DisruptionTarget",
                "status": "True",
                "lastTransitionTime": "2023-05-23T17:00:55Z",
                "reason": "DeletionByPodGC",
                "message": "PodGC: node no longer exists",
              },
            ],
        },
    },
  "responseObject":
    {
      "kind": "Status",
      "apiVersion": "v1",
      "metadata": {},
      "status": "Failure",
      "message": "failed to create manager for existing fields: failed to convert new object (app-a/app-a-7b7ddc5874-c85hq; /v1, Kind=Pod) to smd typed: errors:\n .spec.containers[name=\"app-a\"].env: duplicate entries for key [name=\"RABBITMQ_HOST\"]\n .spec.containers[name=\"app-a\"].env: duplicate entries for key [name=\"RABBITMQ_PORT\"]\n .spec.initContainers[name=\"db-migration\"].env: duplicate entries for key [name=\"RABBITMQ_HOST\"]\n .spec.initContainers[name=\"db-migration\"].env: duplicate entries for key [name=\"RABBITMQ_PORT\"]",
      "code": 500,
    },
  "requestReceivedTimestamp": "2023-05-23T17:00:55.632119Z",
  "stageTimestamp": "2023-05-23T17:00:55.637338Z",
  "annotations":
    {
      "authorization.k8s.io/decision": "allow",
      "authorization.k8s.io/reason": 'RBAC: allowed by ClusterRoleBinding "system:controller:pod-garbage-collector" of ClusterRole "system:controller:pod-garbage-collector" to ServiceAccount "pod-garbage-collector/kube-system"',
    },
}
```

### Kubernetes version

<details>

```console
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.1", GitCommit:"e4d4e1ab7cf1bf15273ef97303551b279f0920a9", GitTreeState:"clean", BuildDate:"2022-09-14T19:49:27Z", GoVersion:"go1.19.1", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"26+", GitVersion:"v1.26.4-eks-0a21954", GitCommit:"4a3479673cb6d9b63f1c69a67b57de30a4d9b781", GitTreeState:"clean", BuildDate:"2023-04-15T00:33:09Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
AWS EKS
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Amazon Linux"
VERSION="2"
ID="amzn"
ID_LIKE="centos rhel fedora"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2"
HOME_URL="https://amazonlinux.com/"

$ uname -a
Linux ip-x-x-x-x.region.compute.internal 5.15.108 #1 SMP Tue May 9 23:54:26 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

```

</details>

### Install tools

<details>
Cluster Autoscaler
</details>

### Container runtime (CRI) and version (if applicable)

<details>
ContainerD
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
CNI: Cilium 1.11
CSI: AWS EBS CSI
</details>

**Comment by carlosjgp** [comment]
I found references to this issue here
- https://github.com/kubernetes/kubernetes/pull/113245#discussion_r1002987037
- https://bugzilla.redhat.com/show_bug.cgi?id=1945548

**Comment by carlosjgp** [comment]
Log statement "failed to create manager..." found in this fuction
https://github.com/kubernetes/kubernetes/blob/2e306e9632b13a3c1530a7e7ddf52a50f4741523/staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/skipnonapplied.go#L75-L89

**Comment by aojea** [comment]
> I found references to this issue here
> 
> * [add warning for dup ports in containers[*].ports and service.ports #113245 (comment)](https://github.com/kubernetes/kubernetes/pull/113245#discussion_r1002987037)
> * https://bugzilla.redhat.com/show_bug.cgi?id=1945548

yep https://github.com/kubernetes/kubernetes/pull/113245

/cc @apelisse @thockin @pacoxu

**Comment by pacoxu** [comment]
I just update the PR https://github.com/kubernetes/kubernetes/pull/113245#issuecomment-1567749104.

**Comment by carlosjgp** [comment]
> I just update the PR [#113245 (comment)](https://github.com/kubernetes/kubernetes/pull/113245#issuecomment-1567749104) and the PR is for the pod warning part.
> 
> For service, there is a validation error for dup port+protocol, and I tried to add a warning for dup ports(different protocol) in #113245.
> 
> There is a thread in [#113245 (comment)](https://github.com/kubernetes/kubernetes/pull/113245#discussion_r1002006603).

Do notice that the reason to open this bug is the fact that duplicated ports on PODs (containerPort+protocol) or environment variables seem to break the Pod Garbage collector that runs on the Kubernetes Control Plane which appears to be a side effect of the discussion on the PR mentioned.

But the case I'm experiencing is not covered in the comments of that PR I'll chip in into that conversation

**Comment by SergeyKanzhelev** [comment]
/remove-sig node
/sig network

for duplicated pods

**Comment by aojea** [comment]
@mimowo the problem seems to come because of this code

https://github.com/kubernetes/kubernetes/blob/04fcbd721cd3e72eb9375275891e9bc3e28c3341/pkg/controller/podgc/gc_controller.go#L325-L343

since the feature gate is enabled by default in 1.26, when the ServiceSideApply fails the pod is never deleted.

I think we should delete the pod despite the service side error to keep the backwards compatible behavior, and the SSA error should not block the deletion

**Comment by mimowo** [comment]
In pod failure policy (and disruption conditions for that matter) it was a deliberate decision to not proceed with DELETE if the PATCH fails. It could be considered a bug if the condition isn’t added, given the [KEP Design Details](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures#new-podconditions)), and the [user-facing docs](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-conditions).  Not adding the condition would also undermine the usefulness of the pod failure policy.

When the PATCH request fails, the controller, such as PodGC, will retry. If the reason for the failure is transient it would succeed next time (ofc transient failures could also happen to the DELETE request itself).

I think the issue is that the PATCH fails permanently due to inconsistency between validation (allowing for duplicated env. vars), and what SSA can handle (which cannot process requests with the duplicated entries). Thus, IIUC, the bug in SSA or the validation code. IMO, this is a validation issue, because probably duplicated entries user's mistake, that is just not picked up by validation. Still, maybe it could be supported by SSA nevertheless.

If this issue cannot be resolved at the SSA level or validation, then I see three options:
1. Accept as known issue (add to known issues( and improve documentation (add warning to using PodDisruptionConditions)
2. Replace SSA with regular patches, while against the long-term plan of k8s migration to SSA, it would solve this scenario for a time-being
3. Introduce a knob (either in control plane or per-pod) to proceed with delete in case of patch failing (potentially after a couple of retries, so that the condition isn’t dropped in case of a single dropped PATCH request), but this may result in worst throughput and other user complains, if the root cause isn’t fixed.

/cc @alculquicondor

**Comment by alculquicondor** [comment]
This error message suggests that you have a buggy webhook trying to modify the Pod object on status updates.

```
failed to create manager for existing fields: failed to convert new object (app-b/app-b-5894548cb-7tssd; /v1, Kind=Pod) to smd typed: .spec.containers[name="app-b"].ports: duplicate entries for key [containerPort=8082,protocol="TCP"]
```

Could you verify this @carlosjgp?

**Comment by alculquicondor** [comment]
Oh sorry, I missed the fact that your pod is already violating a validation rule from the beginning.
But then, how was the pod created?
Did we tightened validation at some point? @apelisse

**Comment by alculquicondor** [comment]
Looking at the current validation code (as of v1.27), you shouldn't be able to create these Pods. I'm trying to check whether this validation was added in v1.25 or v1.26, but it looks like it was there before.

**Comment by alculquicondor** [comment]
Summary of current problem.

- The API validation doesn't check for duplicated ports or env variables. Thus, pods can be created with this violation.
- We added the port numbers and environment variables as mapKeys in the API https://github.com/kubernetes/kubernetes/pull/70100. As a result, server-side-apply cannot succeed with these pods.

Thanks to @aojea for pointing this out.

In general, this can be considered a bug in the pod validation, however, adding such validation can break some users. Still, we can try to prevent new pods like this from being created in future versions, to limit future impact (in addition to the warning). I'll follow up on this in https://github.com/kubernetes/kubernetes/pull/113245.

The next thing we can do is something similar to option 3 by @mimowo:
- Detect the type of error we get from the apiserver on SSA. If it's the duplicated key error, then proceed with the deletion without adding the pod condition.

**Comment by alculquicondor** [comment]
> In general, this can be considered a bug in the pod validation

This is already being tracked here: https://github.com/kubernetes/kubernetes/issues/113482

**Comment by mimowo** [comment]
> Detect the type of error we get from the apiserver on SSA. If it's the duplicated key error, then proceed with the deletion without adding the pod condition.

Yeah, but then what about other controllers that would use SSA for patching pods (except for k8s built in controllers also some other projects may be using SSA already, such as Kueue). Plumbing them all would proliferate the workaround. 

Is there a chance to support SSA for that pods?

**Comment by alculquicondor** [comment]
That's beyond the scope of this issue.

**Comment by mimowo** [comment]
> That's beyond the scope of this issue.

So you suggest just plumbing PodGC? I could propose a PR "fix", but I'm not convinced this is the way to go about it. We also use SSA to add the condition for api eviction, preemption, etc.

**Comment by alculquicondor** [comment]
We should fix all k8s controllers that use SSA, for the purpose of this issue.

**Comment by mimowo** [comment]
Ok, with this approach we will retry adding the condition on transient failures.

**Comment by mimowo** [comment]
Started the WIP PR to silence the error from PodGC: https://github.com/kubernetes/kubernetes/pull/118398. Will cover more components if we have a consensus to go this way from api-reviewers. I would also appretiate some help with api server work to bubble up the error as a reason.

**Comment by thockin** [comment]
@apelisse  and I are discussing in realtime.

Current thinking:

1) It's valid to have `containerPorts[]` with same `port`+`protocol` and different `hostPort` or `hostIP`.  In fact, it is (unfortunately) allowed to have the same `hostPort` and `hostIP` as long as they are unspecified (the Go zero values).  The SSA keys are wrong.  The current validation is overly loose, which practically means this field can not be treated as a map (or we break users who do this weird-but-valid thing).

2) An env var value can ref other env vars (including a previous instance of its own name!!).  Ergo, it's really not a map at all.

If I change both of these to `+listType=atomic` then SSA works for things which have PodSpec embedded (Pods, Deployments, etc).  @apelisse points out that this is an unfortunate tradeoff - if any client actually wants to co-own container ports (unlikely) or env vars (somewhat more likely), they have to know that this field is atomic, and do a read-modify-write.

I have a branch which implements this.  @apelisse is thinking about how to convince ourselves it is OK.

Also:  https://github.com/kubernetes/kubernetes/issues/118418  and   https://github.com/kubernetes/kubernetes/issues/118417 are some followup.

My branch: https://github.com/kubernetes/kubernetes/compare/master...thockin:kubernetes:ssa_hack_env_ports?expand=1

**Comment by jpbetz** [comment]
As potential band-aid that might be feasible to backport, we could fence off the problematic failure cases, and allow benign server side applies through.

I'm imagining this being a rule like: Allow apply operations for apply configurations that includes entries of a listType=map with duplicates when (a) the apply configurations keys do not intersect with any duplicated keys and (b) the caller does not have field ownership of the duplicated key entries

**Comment by mimowo** [comment]
> I'm imagining this being a rule like: Allow apply operations to listType=maps with duplicates when (a) the apply configuration keys do not intersect with any duplicated keys and (b) the field manager does not have field ownership of the duplicated keys.
> This doesn't fix the problem, but it dramatically narrows down which server side apply requests will be impacted by this.

IIUC this would solve the issue in PodGC and other k8s controllers using SSA for updating the `status` sub-resource. If the scope of the change is smaller compared to the proper "fix" for SSA, then maybe it is a good mid-term alternative, that could be cherry-picked to 1.26. wdyt @apelisse ?

**Comment by thockin** [comment]
Yeah, whatever is the most surgical fix will be best. Changing to atomic
seems to work (?) but it's hard to reason about all the possible corner
cases.

**Comment by apelisse** [comment]
Yes, I think addressing the exact concern is going to help a lot more than changing pods (and all types using podtemplates).
I'll try to give it a try and see how it goes.

**Comment by thockin** [comment]
FWIW - changin to atomic passed CI

https://github.com/kubernetes/kubernetes/pull/118475

**Comment by carlosjgp** [comment]
For now, we are mitigating this with a Cronjob that runs every 10m with this Python code

`requirements.txt`
```
kubernetes==26.1.0 ; python_version >= "3.11" and python_version < "4.0"
loguru==0.6.0 ; python_version >= "3.11" and python_version < "4.0"
```

`garbage_collector__init__.py` empty file

`garbage_collector/__main__.py`
```
from garbage_collector.kubernetes_client import get_core_v1_api, get_orphan_pods, kill_pods

if __name__ == "__main__":
    v1 = get_core_v1_api()
    pods = get_orphan_pods(v1)
    kill_pods(v1, pods)
```

`garbage_collector/kubernetes_client.py`
```
from typing import List

from kubernetes import client, config
from kubernetes.client.models import V1NodeList, V1Pod, V1PodList
from kubernetes.client.rest import ApiException
from loguru import logger


def get_core_v1_api() -> client.CoreV1Api:
    config.load_incluster_config()
    return client.CoreV1Api()


def get_orphan_pods(v1_client: client.CoreV1Api) -> List[V1Pod]:
    """
    Pods that seem to be running in nodes that don't exist anymore
    """
    nodes: V1NodeList = v1_client.list_node(watch=False)
    all_node_names = [no.metadata.name for no in nodes.items]
    logger.info(f"Nodes found: {all_node_names}")
    pods: V1PodList = v1_client.list_pod_for_all_namespaces(watch=False)
    return [po for po in pods.items if po.spec.node_name not in all_node_names]


def kill_pods(v1_client: client.CoreV1Api, pods: List[V1Pod]) -> None:
    logger.info(f"Orphan PODs found: {len(pods)}")
    for pod_to_kill in pods:
        # Create an instance of the API class
        name = pod_to_kill.metadata.name
        namespace = pod_to_kill.metadata.namespace

        logger.info(f"Killing {namespace}/{name}")
        try:
            v1_client.delete_namespaced_pod(name, namespace, grace_period_seconds=0)
            logger.info(f"Killed {namespace}/{name}")
        except ApiException as k8s_exception:
            logger.warning(
                f"Exception when calling CoreV1Api->delete_namespaced_pod: {k8s_exception}",
            )
```

**Comment by thockin** [comment]
Given how deep this runs, are we going to back-out SSA from controllers?

**Comment by alculquicondor** [comment]
The plan is to do a try with SSA and retry with a regular status update #118398 


FYI on https://github.com/kubernetes/kubernetes/issues/118261#issuecomment-1580530296: it won't work for Jobs

**Comment by mimowo** [comment]
@apelisse @thockin what about the @jpbetz proposal https://github.com/kubernetes/kubernetes/issues/118261#issuecomment-1574200882, would it work given that the controllers patch status, and the conflicting fields are in spec? This seems to me like a smaller change than patching all controllers.

**Comment by apelisse** [comment]
That's what we're going for. It's not totally trivial to implement but I'll make sure it lands in 1.28 (and ideally can be backported further back).

**Comment by thockin** [comment]
I still think we need to consider changing to atomic for some of these, or an `appendOnly` apply semantic (which seems unlikely) or a multi-map (map[key]set[value]) or something.

Env is literally NOT a map, it should probably be atomic.

Would we consider an SSA option that says "if you hit dup keys, just pick one and drop the others" ?  More like client-side?  If we did that we could  maybe make some of the errant types have all of their fields as keys (e.g. for ContainerPorts exact dups are semantically identical)

**Comment by mimowo** [comment]
IIUC these are the options for the short-term (cherry-pick down to 1.26) solution:
1. Only fail SSA when the apply affects the duplicated fields
2. Change to atomic
3. Fallback in all controllers to regular PATCH or PUT

The status: (1.) is currently being investigated; (2.) needs further thought of consequences; (3.) Requires a preparatory solution for https://github.com/kubernetes/kubernetes/issues/118476.

**Comment by thockin** [comment]
That one is known, and covered in https://github.com/kubernetes/kubernetes/pull/118475 - whether `atomic` is the fix we go with is TBD, but that field is known bad :)

**Comment by mimowo** [comment]
Bumping the thread with cross referencing another issue report on that: https://github.com/kubernetes/kubernetes/issues/118261. 

@kubernetes/sig-api-machinery-members  we would like to get input on the preferred way of fixing this that could be cherry-picked down to 1.26.

**Comment by thockin** [comment]
You cross-reffed this to itself?

**Comment by alculquicondor** [comment]
#118741 :)

**Comment by thockin** [comment]
xref https://github.com/kubernetes-sigs/structured-merge-diff/issues/234

**Comment by cameronattard** [comment]
Any idea when a fix will be available on 1.26?

**Comment by mimowo** [comment]
> Any idea when a fix will be available on 1.26?

Good question. I was hoping for a API server side fix as in [comment](https://github.com/kubernetes/kubernetes/issues/118261#issuecomment-1584136887), which was admitted as a way to try in [comment](https://github.com/kubernetes/kubernetes/issues/118261#issuecomment-1584807362). @apelisse any progress on this? 

Another possibility is to skip adding the condition selectively on this error. Attempted in https://github.com/kubernetes/kubernetes/pull/118398, but for this we also need API server support to return a dedicated error as pointed out in the [comment](https://github.com/kubernetes/kubernetes/pull/118398#issuecomment-1682482206). I think this solution is less preferred anyway as some features not only depend on the extra condition, but also on the fact that the pods are transitioned to terminal phase - PodGC uses SSA to also set pod phase to `Failed`. From the top of my head features that rely on this: [pod failure policy](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures) and [pod replacement policy](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3939-allow-replacement-when-fully-terminated), but over time there might be more.

@liggitt any recommendation how to proceed?

**Comment by apelisse** [comment]
I'm working on this right now, the change is full of subtle nuances that I'm trying to deal with right now. I think I should be done soon, definitely will be merged for the next release, hopefully with a lot of time left for you to address the bugs its causing.

**Comment by korjek** [comment]
we've created a cronjob that deletes pods that are stuck in `Terminating` (such pods have `deletion_timestamp` set in `metadata`) to get cluster cleaner.
But the issue that impacts us is pods that suck in `Running` state while the node has been removed. 
In this case, deployment is not able to spinup a new replica. 
Any recommendations on how to deal with/discover such pods?

**Comment by liggitt** [comment]
> @liggitt any recommendation how to proceed?

Sorry, I just caught up here.

The SSA fix for this is likely to be complex / subtle as @apelisse noted, and is not something I think we should backport to 1.26/1.27/1.28 (which is where this issue is surfacing).

At first glance, I would do one of the following for 1.26/1.27/1.28:
* disable the PodDisruptionConditions feature gate by default
* make the things adding the condition proceed deleting even if it can't add the condition using SSA
* switch the things adding the condition to use another method (update or jsonpatch or something) instead of SSA

**Comment by alculquicondor** [comment]
Let's go with:
> switch the things adding the condition to use another method (update or jsonpatch or something) instead of SSA

/unassign @thockin 
/assign @mimowo

**Comment by mimowo** [comment]
I have opened a PR to fix this issue for PodGC: https://github.com/kubernetes/kubernetes/pull/121103.

**Comment by sfotony** [comment]
Ticket opened with KD team [here](https://jira.corp.adobe.com/browse/ETHOS-52146).

**Comment by jgoeres** [comment]
We recently ran into this issue with a pod that has a duplicate environment variable in one of our first 1.27 test servers (we have various ways to "calculate" or allow users to customize setting env vars in our Helm chart). We shut down the nodes over night to reduce cost. In the morning, we found those pods in Terminating state. The API server logs indicated the root cause and brought us here. 
Aside from the actual bug, I think the handling of environment variables should be reconsidered. Since from an API perspective they are a list (I guess there was a reason for this when the podspec was designed), but in the end they become a map, there are basically two options to fix this: always raise an error when env vars (or ports or any other "this could have been a map" list entries) are duplicated, forcing users to resolve the issue, not ignore it like we did (and we didn't even do so consciously - Gitops tools don't care about warnings)  OR at least have a well-defined order of precedence (e.g., "last specified value wins")

**Comment by alculquicondor** [comment]
What you suggest is tracked in #113482, but the solution is not trivial.

**Comment by alculquicondor** [comment]
yes, if we were to have k8s 2.0, we might go that path. But in 1.x, we need to keep backwards compatibility. So we need to fix how SSA validates the keys.

**Comment by liggitt** [comment]
Agreed. In general, our compatibility stance is that if a manifest resulted resulted in a successfully running pod in a previous release, we will only emit warnings about potential issues, rather than rejecting the request in a new release.

**Comment by thockin** [comment]
> In general, our compatibility stance is that if a manifest resulted resulted in a successfully running pod in a previous release, we will only emit warnings about potential issues, rather than rejecting the request in a new release.

FWIW - I think we should consider versioned validation which does reject bad things in new API versions.  "If you NEED them, use the old API" ?  I believe we should do a core/v2 to capture better defaults and to allow some small amount of API reorganization, but I am not ready to pursue that yet (lack of bandwidth).

**Comment by liggitt** [comment]
> FWIW - I think we should consider versioned validation which does reject bad things in new API versions

We've definitely done this in the past, though usually on beta → v1 boundaries.

> I believe we should do a core/v2 to capture better defaults and to allow some small amount of API reorganization, but I am not ready to pursue that yet (lack of bandwidth).

Oof... groupless v1 was treated really specially...

https://github.com/kubernetes/kubernetes/blob/2c10d9cacb0626219753a3df28a76b6bfae2b34a/staging/src/k8s.io/client-go/discovery/discovery_client.go#L379-L383

A grouped v2 would mean permanent appearance of core types both in v1 and in $group/v2. That was a pain for generic discovery-driven controllers when we were migrating things out of extensions/v1beta1 :-/

**Comment by thockin** [comment]
Could we not do a core/v2 in the same manner as v1?  I don't have viz ito the machinery here.

**Comment by mimowo** [comment]
As the issue is fixed and the cherry-picks PR are prepared I would suggest to continue the discussion on the underlying open SSA issue:  https://github.com/kubernetes/kubernetes/issues/113482

**Comment by liggitt** [comment]
fixed are cherry-picked, will be in 1.28.4 / 1.27.8 / 1.26.11
